ONSdigital/census-rh-ui
# CR-1456 Add log identifier

# Motivation and Context
Enable logs to be filtered to associate previous log statements generated from an individual client and provide context in investigating client errors, warnings.

# What has changed
Middleware has been introduced so that for every request a client_id and trace are added to the request object and available to log statements.
The trace is taken from the X-Cloud-Trace-Context header which is added to a request by Google and has the specification `X-Cloud-Trace-Context: TRACE_ID/SPAN_ID;o=TRACE_TRUE`. A trace is unique to a client request and is used to group logs within a client request process. A span would represent a unit of work within a trace, if for instance we were using OpenTelemetry or OpenCensus client libraries to instrument the code and generate logs and metrics and display latency data in Cloud Trace. Using these could have performance implications in production and not to be considered at this stage, so I have simply added the trace to log statements to group request logs and ignored span as unnecessary noise with no purpose for production.
The client_id is used to group all requests from a client during a session. It is stored within a session and made available by adding to the incoming request if available in the session, or created if not.
I have also taken the opportunity to make the client_ip available in the same manner, removing the frequent calls to `self.setup_request(request)` in each view handler.

# How to test?
New unit tests added. Run against cucumber tests. I have currently deployed this code to the development environment so you can generate log statements and see the client_id and trace in the jsonPayload data. 

# Links
https://collaborate2.ons.gov.uk/jira/browse/CR-1456
